### PR TITLE
Cast PyArrow schema to `large_*` types

### LINKED ISSUE #791: Upcasting and Downcasting inconsistencies with PyArrow Schema
### Apache Iceberg version

0.6.0 (latest release)

### Please describe the bug 🐞

`schema_to_pyarrow` converts BinaryType to `pa.large_binary()` type. This creates inconsistencies with the arrow table schema produced from the data scan between:
1. when schema_to_pyarrow is used when there is no data in the table (pa.large_binary())
2. when we use the physical_schema of the file fragment to read the table (pa.binary())

Related PR: https://github.com/apache/iceberg-python/pull/409 

The implication of this bug is that pa.Table read from the same Iceberg Table may yield different schema based on whether or not there is data within the defined table scan.

More importantly, it also means that if one of the files is empty, and another file has data within the same table scan, then the schema inconsistencies in the two arrow tables will result in an error as we attempt to `pa.concat_tables(tables)`

### PR DESCRIPTION
**Pull request by syun64**
Fixes #791 

For consistency, we should always cast to large types when inferring pyarrow schema from Iceberg Schema, and when scanning using the physical_schema of the fragment

### CODE REVIEW

**Review by HonahX — COMMENTED** [review]
Thanks for working on this! It looks great. Just have 2 minor comments

To summarize the discussion in https://github.com/apache/iceberg-python/issues/791, we could always benefit from reading data as `large_*` type since offset is 64-bit. For parquet, we will still write data in non large type due to parquet's 2GB data size limitation. 

Just to confirm my understanding, since the major difference between `large_binary` and `binary` is the offset type (64-bit versus 32-bit), there will be no significant increase in memory usage when reading data as `large_binary`.

File: pyiceberg/io/pyarrow.py [hunk 1]
```diff
@@ -680,6 +680,10 @@ def _pyarrow_to_schema_without_ids(schema: pa.Schema) -> Schema:
     return visit_pyarrow(schema, _ConvertToIcebergWithoutIDs())
 
 
+def _pyarrow_with_large_types(schema: pa.Schema) -> pa.Schema:
```
[the remark this answers is not in the document]
Comment (syun64) [reply pyiceberg/io/pyarrow.py:683]: I really like this name suggestion - thank you!

**Review by HonahX — APPROVED** [review]
Thanks for the detailed explanation!

**Comment by syun64** [comment]
@HonahX could I ask for you to merge this in? It'll help unblock me in https://github.com/apache/iceberg-python/pull/786/files

**Review by Fokko — APPROVED** [review]
This looks good, thanks @syun64 for adding this, and @HonahX for the review 👍 

One possible addition would be to have this configurable. Based on configuration, it would either go with normal or with large types.

**Pull request merged**